internal/campaigns/execution_cache.go

```diff
@@ -77,58 +77,91 @@ func (c ExecutionDiskCache) cacheFilePath(key ExecutionCacheKey) (string, error)
 		return "", errors.Wrap(err, "calculating execution cache key")
 	}
 
-	return filepath.Join(c.Dir, keyString+".diff"), nil
+	return filepath.Join(c.Dir, keyString+".v3.json"), nil
 }
 
-func (c ExecutionDiskCache) Get(ctx context.Context, key ExecutionCacheKey) (string, bool, error) {
+func (c ExecutionDiskCache) Get(ctx context.Context, key ExecutionCacheKey) (ExecutionResult, bool, error) {
+	var result ExecutionResult
+
 	path, err := c.cacheFilePath(key)
 	if err != nil {
-		return "", false, err
+		return result, false, err
 	}
 
 	data, err := ioutil.ReadFile(path)
 	if err != nil {
 		if os.IsNotExist(err) {
 			err = nil // treat as not-found
 		}
-		return "", false, err
+		return result, false, err
 	}
 
-	// We previously cached complete ChangesetSpecs instead of just the diffs.
-	// To be backwards compatible, we keep reading these:
-	if strings.HasSuffix(path, ".json") {
-		var result ChangesetSpec
+	// There are now three different cache versions out in the wild and to be
+	// backwards compatible we read all of them.
+	// 2-step plan for taming this:
+	//  1) February 2021: deprecate old caches by deleting the files when
+	//                    detected and reporting as cache-miss.
+	//  2) May 2021 (two releases later): remove handling of these files from
+	//                                    this function
```

Comment:
:+1:, particularly since the worst case here is that some campaign steps have to be re-executed.

Can you please open a GitHub issue to cover the follow up action, assuming you haven't already?

Reply:
Done: https://github.com/sourcegraph/sourcegraph/issues/17306